### ISSUE TITLE
Error importing ES6 moules

### ISSUE THREAD

**Issue by Lock-up** [Flyer53/jsPanel4#74]
Hi,
after updating from 4.4.0 to 4.5.0 I get following error:

    jspanel.js:3974 Uncaught TypeError: Cannot assign to read only property 'exports' of object '[object Object]'
        at Module.<anonymous> (jspanel.js:3974)
        at Module../node_modules/jspanel4/es6module/jspanel.js (jspanel.js:3975)
        at __webpack_require__ (bootstrap:19)
        at Module../src/jsPanel.config.ts (jsPanel.config.ts:1)
        at __webpack_require__ (bootstrap:19)
        at Module../src/main.ts (main.ts:1)
        at __webpack_require__ (bootstrap:19)
        at Object.0 (main.ts:21)
        at __webpack_require__ (bootstrap:19)
        at bootstrap:83

File import:

    import { jsPanel } from 'jspanel4/es6module/jspanel.js';
    import 'jspanel4/es6module/extensions/tooltip/jspanel.tooltip.js';

Problem seems to be the CommonJS module export:

    // Add CommonJS module exports, so it can be imported using require() in Node.js
    // https://nodejs.org/docs/latest/api/modules.html
    if (typeof module !== 'undefined') {
        module.exports = jsPanel;
    }

I think this SO threads are related to this:
https://stackoverflow.com/questions/42449999/webpack-import-module-exports-in-the-same-module-caused-error
https://stackoverflow.com/questions/31354559/using-node-js-require-vs-es6-import-export

if I got it right, mixing import/export and require()/module.exports = ... is not allowed.

Greetings Jan

**Comment by Flyer53** [comment]
@Lock-up 
Thanks for the infos.
I need to look into this myself before I can say anything useful ;)

Did you check whether it works for you when you remove the CommonJS exports?

**Comment by Lock-up** [comment]
Hi, yes it's working if I remove the exports.
________________________________

**Comment by Flyer53** [comment]
@Lock-up 
I personally never used anything else than ES6 modules. And since it's a standardized javascript feature now I don't see a reason to make it more complicated.

You seem to have found how to easily make it work in your environment. And as far as I read you it's not a jsPanel bug ... just an adaption that might be required depending on the "type" of modules used.

As far as I know node is also working on the implementation of native javascript modules.

Bottom line: I don't see a reason to make any changes concerning the use of non-standard modules.

**Comment by Lock-up** [comment]
@Flyer53 
Hi,
in my opinion it is not the standard es6 way of exporting modules:
https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Statements/export

It is the es5 where two module systems exist:
http://exploringjs.com/es6/ch_modules.html#sec_modules-in-javascript

Am I right? or did I miss something?

**Comment by Flyer53** [comment]
@Lock-up 
Hi again,
Well, I don't consider myself a professional programmer. Here is how I understand it:
Strictly spoken there is no such thing as "real" modules in ES5. CommonJS (node) and AMD are simply javascript libraries that somehow implement/mimic a feature/behaviour you would expect from a "real" module in order to benefit from some advantages "real" module provide.

With ES6 came real modules and that's what I want to support with the files provided in the `es6module` folder. And since it works just fine I don't think I made something wrong here. (MDN is my primary source to learn about certain js features btw.)
The very last 3 lines of code in the files don't have anything to do with ES6 modules and are simply added to make the files work with node.js. Those 3 lines were added upon request of another user (see #61) who needed them to work with node and I assume that it works since the PR came from him and he hopefully tested it thoroughly before making the PR.

Beyond that jsPanel has no built-in support for AMD. If you need that you have to implement it yourself.

As I said before, I'm not a professional programmer and I don't have a single reason to learn CommonJS or AMD with native modules available and supported by all major browsers.

That you can't mix the various types of module implementations seems quite natural to me.

And if I got you right it's working for you as well when you simply remove the last 3 lines of code.

If you have a single solution that works with all 3 module environments I'd be happy if you share it with me 😄

**Comment by Lock-up** [comment]
In my opinion the last 3 lines shouldn't be in the code for the `es6module` folder because they are not the way es6 handles modules.

I don't know how you create the different versions `dist` and `es6module` if the files in `dist` are <=es5 I think it is okay that they include the CommonJs module syntax, but the files in `es6module` shouldn't.

Because as you said:

> That you can't mix the various types of module implementations seems quite natural to me.

I would suggest es6module should handle import/export only in the way it is designed in es6.

So you have to create 2 different versions, what you are already doing, don't you?

If you are using babel, maybe this helps?
https://gitlab.com/sebdeckers/babel-plugin-transform-commonjs-es2015-modules

**Comment by Flyer53** [comment]
>In my opinion the last 3 lines shouldn't be in the code for the es6module folder because they are not the way es6 handles modules.

There you have a point ...

And thanks for the link, I'll take a look...

**Comment by Flyer53** [comment]
@Lock-up 
For future releases I'll try to remember to remove those last 3 lines of code in the `es6module` folder in order to prevent this error msg.

In case you don't have anything else concerning this issue I'd ask you to close it.

Thanks,
Stefan

**Comment by Lock-up** [comment]
OK, thank you for your support!

**Comment by vlad-saritasa** [comment]
Why this issue is closed? I have the same bug using npm & webpack. The default configuration is not working.

**Comment by Flyer53** [comment]
@vlad-saritasa 
Once version 4.7.0 is released all files in the `es6module` folder will be pure ES6 modules without any other disturbing code.
You can [download the current v4.7.0-alpha.20](https://alpha.jspanel.de/downloads/jspanel-4.7.0-alpha.20.zip) to check it out. But it's not yet available via GitHub or NPM.
